### ISSUE TITLE
Allow class-scoped framing

### ISSUE THREAD

**Issue by azaroth42** [w3c/json-ld-framing#29]
It would be useful to allow individual classes to have their own frames, such that a class which can appear at many points in the structure can be defined only once and then the definition inherited.

For example, a class like Type / Concept / Tag might appear anywhere in a structure associated with any other class. Currently (I believe) you would need to define all possible structures to ensure that the Type was framed consistently.

**Comment by gkellogg** [comment]
What do you think this might look like? Framing serves as a "program by example", and I'm not sure what such a frame might look like.

**Comment by azaroth42** [comment]
I can't recall what the issue was, closing :)

**Comment by azaroth42** [comment]
Ahha, and in coming back to an implementation, I recall the issue. It's related to https://github.com/w3c/json-ld-syntax/issues/19 and https://github.com/w3c/json-ld-api/issues/33.

Framing is a program-by-example or templating system, but it requires the absolute structure to be known from the root node down. We have a "Type" (or Concept, or Tag) that has a structure we'd like to see repeated every time it is referenced, but don't know all of the places in the structure where it might occur.

So we'd like the following to be able to be specified in a frame, without knowing the structure for `property`, just to make a frame for resources that have a `type` of `Tag`:

```
{
  "tag": {
    "id": "http://example.org/tag1",
    "type": "Tag",
    "label": "example tag"
  },
  "property": {
    "more_properties": {
      "tag": {
        "id": "http://example.org/tag1",
        "type": "Tag",
        "label": "example tag"
      }
    }
  }
}
```

Rather than the second instance being just the URI.

**Comment by azaroth42** [comment]
In terms of potential syntax, the best I can come up with is a new keyword (I hear the groans from around the world) that would be a wildcard for properties. It can't be a second graph shape in `@graph` as the processor wouldn't know which shape was the top, and which was the child template. It can't use `@nest` as then you wouldn't be able to frame nested (structural, transparent) properties.

e.g. something like:
```
{
  "@type": "TopLevelClass",
  "@anywhere": [
    { "tag": 
     {
        "@type": {"id": "Tag", "@embed": always"}
        "label": {"@embed": "always"}
      }
    }
  ],
  "property": { ... }
}
```

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-02-15-json-ld#section5).

- `RESOLVED:  Work on a proposal for solving framing#29 along the lines of embedded contexts / scoped contexts, but for embedded sub-frames {: #resolution3 .resolution}`
<details><summary><i class="differentiate">View the transcript</i></summary>
5. Class-scoped Framing (issue frame29)<br/>
Rob Sanderson: <a href="https://github.com/w3c/json-ld-framing/issues/29">https://github.com/w3c/json-ld-framing/issues/29</a><br/>
Rob Sanderson: framing is like programming by example;<br/>
… you need to know the exact structure from the root down.<br/>
… For some properties (like “link” or “tag”),<br/>
Rob Sanderson: <a href="https://github.com/w3c/json-ld-framing/issues/29#issuecomment-463295607">https://github.com/w3c/json-ld-framing/issues/29#issuecomment-463295607</a><br/>
Rob Sanderson: you would need to be able to say “anywhere this property appears, it should conform with this structure”<br/>
… but this is not currently possible.<br/>
Ivan Herman: isn’t it related to the issue I raised recently, while reading the framing document (issue 38)?<br/>
… We can not currently have a “bush-shaped” frame, with several patterns,<br/>
… the first matching one being used.<br/>
… Wouldn’t it solve your issue as well?<br/>
Gregg Kellogg: I think it is possible to have a bush.<br/>
Ivan Herman: not for the pattern itself.<br/>
Dave Longley: this is pretty close: <a href="http://tinyurl.com/y356yzo8">http://tinyurl.com/y356yzo8</a><br/>
Dave Longley: to what Ivan wants – in JSON-LD 1.0 framing<br/>
Gregg Kellogg: I think your particular example could still be solved.<br/>
… I’m not sure this completely addresses what Rob wants.<br/>
… This is more like a “macro” feature.<br/>
… It is somehow similar to scoped contexts… Something like a scoped frame.<br/>
Rob Sanderson: +1 to similarity (but not identity) to scoped contexts<br/>
Dave Longley: I think that what Ivan wants is similar.<br/>
… And I do think that there is a gap in the current framing document.<br/>
… It makes sense for us to create like a ‘type frame’.<br/>
Gregg Kellogg: If we do it for types, we should probably do it for properties, too<br/>
Dave Longley: something like @anywhere makes sense to me as well … defining “subframes” at the top of the frame that get applied when certain types or properties are encountered<br/>
Rob Sanderson: @frame :D<br/>
Gregg Kellogg: we might define something parallel to contexts, that could appear anywhere contexts can appear,<br/>
Proposed resolution: Work on a proposal for solving framing#29 along the lines of embedded contexts / scoped contexts, but for embedded sub-frames (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Dave Longley: +1<br/>
Gregg Kellogg: +1<br/>
Simon Steyskal: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ivan Herman: +1<br/>
David I. Lehn: +1<br/>
David Newbury: +1<br/>
Benjamin Young: +1<br/>
Resolution #3: Work on a proposal for solving framing#29 along the lines of embedded contexts / scoped contexts, but for embedded sub-frames {: #resolution3 .resolution}<br/>
Gregg Kellogg: We should look to ShEx and SHACL for similar patterns we might leverage<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-04-19-json-ld#section4-2).

- `RESOLVED:  defer framing #29 to be discussed for inclusion in a future call`
<details><summary><i class="differentiate">View the transcript</i></summary>
Class-scoped framing<br/>
Rob Sanderson: link: <a href="https://github.com/w3c/json-ld-framing/issues/29">https://github.com/w3c/json-ld-framing/issues/29</a><br/>
Rob Sanderson: The idea here … instead of having framing on only predicates to also have it on classes, whenever you find a resource of type X you use this subframe. But we’d deferred #38 and this seems to be a subtopic of that.<br/>
Gregg Kellogg: Deferred just means we’re not working on this now. Or do you mean defer into another WG?<br/>
Rob Sanderson: Well, defer and we’ll see what happens.<br/>
Ivan Herman: Wait, we’re getting to feature freeze. Whatever is deferred at feature freeze means it goes to another version.<br/>
Gregg Kellogg: I think it meant we weren’t accepting new proposals. It’s still on the docket to work on, but the WG might decide to defer to a later version.<br/>
Ivan Herman: Deferring means that this will not be part of the upcoming version of JSON-LD.<br/>
Gregg Kellogg: Ok, I’m not prepared to say that for this or other things we’ve said “deferred” to.<br/>
Rob Sanderson: I’ve also thought that way, not necessarily closing. If not we’re going to do it we should close.<br/>
… SRI and references to meta data we should continue to discuss it even though they are deferred.<br/>
Ivan Herman: As soon as we are CR we freeze it. For our own time tables we have to have a point where we freeze it for ourselves as well. At that point, if we say “defer” … “if after publishing a new REC, this group or any other group should look at this feature”<br/>
Rob Sanderson: I’m fine with subtly changing the semantics of “defer” when we’re in feature freeze. We should take those things that we’ve marked as defer as “in or out”<br/>
Gregg Kellogg: I think we should decide on the next call specific features we’re going to work on for this version. Otherwise it’s going to be nebulous. It’s premature to say those things that don’t get into this WD are deferred. We need to prioritize issues and set a time by which we will have finalized them.<br/>
Ivan Herman: We are dangerously close to that.<br/>
Rob Sanderson: +1<br/>
Gregg Kellogg: So that’s the topic for the next call.<br/>
Rob Sanderson: How about for now we defer based on #38?<br/>
Ivan Herman: Let’s close the topic for now and then look at all these in general.<br/>
Proposed resolution: defer framing #29 to be discussed for inclusion in a future call (Rob Sanderson)<br/>
Gregg Kellogg: +1<br/>
Ruben Taelman: +1<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Dave Longley: +1<br/>
David I. Lehn: +1<br/>
Resolution #3: defer framing #29 to be discussed for inclusion in a future call<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-04-26-json-ld.html#section3-2).

- `RESOLVED:  Defer framing #29 until future WG`
<details><summary><i class="differentiate">View the transcript</i></summary>
3.2. class-scoped framing<br/>
Ivan Herman: link: <a href="https://github.com/w3c/json-ld-framing/issues/29">https://github.com/w3c/json-ld-framing/issues/29</a><br/>
Rob Sanderson: idea is that framing starts with a tree, but if you don’t know where a class starts in a tree then things get complicated<br/>
… likely would require multiple frames within a frame document<br/>
Benjamin Young: +1<br/>
David Newbury: +1<br/>
Proposed resolution: Defer framing #29 until future WG (Rob Sanderson)<br/>
Ivan Herman: +1<br/>
Dave Longley: +1<br/>
David Newbury: +1<br/>
Ruben Taelman: +1<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
Pierre-Antoine Champin: +1<br/>
David I. Lehn: +1<br/>
Tim Cole: +1<br/>
Simon Steyskal: +1<br/>
Resolution #3: Defer framing #29 until future WG<br/>
</details>

**Comment by AtesComp** [comment]
Since the @graph keyword (and using no @graph keyword) implies that "the absolute structure to be known from the root node down", then maybe using a **@subgraph** keyword would specifies a matching substructure (and might be used multiple times).  Non-matching portions should be included "as is" from the input graph.  This looks like a search and replace operation.
